### PR TITLE
[Clang][Parser] Fix name lookup of template parameters for out-of-line specializations

### LINKED ISSUE #64082: Clang incorrectly parses template type parameter as a concept name
When specializing a class from another namespace, if a type parameter has the same name as a concept in that namespace, Clang mistakenly parses the type parameter name as being the concept name.

For `m` below, the specialization is in the same namespace, the bug does not occur. `MTypeParam` is parsed as the type parameter.

For `n` below, the specialization is outside of the namespace, and `NTypeParam` is treated as the `n::NTypeParam` concept instead of the specialization's `NTypeParam` type parameter in the function parameter list.

https://godbolt.org/z/47ahesrbn

```cpp
template <class T>
concept C = true;

namespace m {

template <class T>
struct M;
template <class T>
concept MTypeParam = true;

template <C MTypeParam>
struct M<MTypeParam> {
    static void f(const MTypeParam&) {}
};

}

namespace n {
template <class T>
struct N;
template <class T>
concept NTypeParam = true;
}

template <C NTypeParam>
struct n::N<NTypeParam> {
    static void f(const NTypeParam&) {}   // error: expected 'auto' or 'decltype(auto)' after concept name
};

int main() {
    m::M<int>::f(2);
    n::N<int>::f(2);
}
```

Broken on Clang 16 and HEAD.
Works on GCC and MSVC.

### PR DESCRIPTION
**Pull request by zyn0217**
Since the implementation of DR458 (d1446017), we have had an algorithm that template parameters would take precedence over its parent scopes at the name lookup. However, we failed to handle the following case where the member function declaration is not yet deferral parsed (This is where the patch of DR458 applies):

```cpp
namespace NS {
  int CC;
  template <typename> struct C;
}

template <typename CC>
struct NS::C {
  void foo(CC);
};
```

When parsing the parameter of the function declaration `void foo(CC)`, we used to perform a name lookup following such a Scope chain:
```
FunctionScope foo (failed)
RecordScope C (failed)
NamespaceScope NS (found `int CC`)
(If failed)
TemplateParameterScope of C
```
This doesn't seem right because according to `[temp.local]`, a template parameter scope should be searched before its parent scope to which the parameter appertains. This patch corrects the search scopes by setting a lookup Entity for template parameter Scopes so that we can bail out in CppNameLookup() when reaching the RecordScope. Afterward, the search chain would be like:
```
FunctionScope foo (failed)
RecordScope C (failed)
TemplateParameterScope of C (found CC)
(If failed)
NamespaceScope NS
```

Fixes https://github.com/llvm/llvm-project/issues/64082

### CODE REVIEW

**Comment by zyn0217** [comment]
Gently Ping

File: clang/test/CXX/temp/temp.res/temp.local/p8.cpp [hunk 1]
```diff
@@ -151,4 +152,25 @@ namespace SearchClassBetweenTemplateParameterLists {
...
+  void foo(CC) {} // This should find the template type parameter.
+};
+
+#if __cplusplus >= 202003L
+
+template <True CC>
+struct SearchClassBetweenTemplateParameterLists::D<CC> {
+  void foo(CC);
+};
+
```
Comment (cor3ntin) [inline clang/test/CXX/temp/temp.res/temp.local/p8.cpp:175]: - Can you add a test with NTTP template params?
- Do we have tests for that sort of lookup in requires clause / return types?
- Maybe a test that check that if both names are a `typename` we find the correct one

**Review by cor3ntin — COMMENTED** [review]
Thanks for the detailed description!
I think the change makes sense, we just might need more tests and a changelog entry

**Comment by cor3ntin** [comment]
I am sorry this fell off my radar. Feel free to ping weekly in the future

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by zyn0217** [comment]
Thanks for the review :)

**Pull request merged**